src/pmdas/llc/llc.c

```diff
@@ -0,0 +1,299 @@
+#include "domain.h"
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <dirent.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+/*
+ * LLC  Last-Level Cache of CPU info.
+ *
+ *  Linux supports resctrl which can be used to
+ *  control QoS of CPU cache and monitor LLC(last level cache).
+ *  This feature depends on the 'resctrl' pseudo filesystem.
+ *  The kernel should be built with the relevant config and
+ *  the pseudo-filesystem needs to be mounted:
+ *  mount -t resctrl resctrl -o mba_MBps /sys/fs/resctrl (on Intel)
+ *  mount -t resctrl resctrl -o cdp      /sys/fs/resctrl (on AMD)
+ *
+ *  Which results in directories like:
+ *   /sys/fs/resctrl/mon_data/mon_L3_XX
+ *   
+ *  These counters can be found in doc:
+ *   https://github.com/torvalds/linux/blob/master/Documentation/x86/resctrl.rst
+ *
+ * Metrics
+ *   llc.occupancy % of LLC in use
+ *   llc.mbm_local local memory bandwidth of this LLC
+ *   llc.mbm_total total memory bandwidth of this LLC
+ *
+ */
+
+/*
+ * instance domains
+ */
+
+const int LLC_INDOM = 0;
+
+static pmdaIndom indomtab[] = {
+    { LLC_INDOM, 0, NULL }
+};
+
+/*  occupancy   LLC:0:0
+    mbm_local   LLC:0:1
+    mbm_total   LLC:0:2   */
+
+static pmInDom	*llc_indom = &indomtab[LLC_INDOM].it_indom;
+
+static pmdaMetric metrictab[] = {
+    // occupancy
+    { NULL, 
+      { PMDA_PMID(0,0), PM_TYPE_FLOAT, LLC_INDOM, PM_SEM_INSTANT, 
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+    // mbm_local
+    { NULL, 
+      { PMDA_PMID(0,1), PM_TYPE_U32, LLC_INDOM, PM_SEM_INSTANT, 
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+    // mbm_total
+    { NULL, 
+      { PMDA_PMID(0,2), PM_TYPE_U32, LLC_INDOM, PM_SEM_INSTANT, 
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+};
+
+static int	isDSO = 1;		/* =0 I am a daemon */
+static char	*username;
+static char	mypath[MAXPATHLEN];
+static char	llcdir[MAXPATHLEN];
+
+typedef struct
+{
+    float llc_occupancy;
+    unsigned long llc_mbm_local;
+    unsigned long llc_mbm_total;
+} llc_metrics;
+    
+/* command line option handling - both short and long options */
+static pmLongOptions longopts[] = {
+    PMDA_OPTIONS_HEADER("Options"),
+    PMOPT_DEBUG,
+    PMDAOPT_DOMAIN,
+    PMDAOPT_LOGFILE,
+    PMOPT_HELP,
+    PMDA_OPTIONS_END
+};
+
+static pmdaOptions     opts = {
+    .short_options = "D:d:l:?",
+    .long_options = longopts,
+};
+
+/*
+ * callback provided to pmdaFetch which handles one metric and one instance
+ */
+static int
+llc_fetchCallBack(pmdaMetric *mdesc, unsigned int inst, pmAtomValue *atom)
+{
+    unsigned int cache_item = pmID_item(mdesc->m_desc.pmid);
+    llc_metrics *lspm;
+    
+    int sts = pmdaCacheLookup(*llc_indom, inst, 0, (void*)&lspm);
+    if (sts < 0 || sts == PMDA_CACHE_INACTIVE)
+	return PM_ERR_INST;
+
+    switch (cache_item) {
+    case 0:		/* occupancy */
+	atom->f = lspm->llc_occupancy;
+	break;
+    case 1:		/* mbm local */
+	atom->ull = lspm->llc_mbm_local;
+	break;
+    case 2:		/* mbm total */
+	atom->ull = lspm->llc_mbm_total;
+    }
```

Comment:
Recommend a 'break' here so it doesn't accidentally flow through to the next persons code addition.